vault/resource_secrets_sync_association.go

```diff
@@ -0,0 +1,201 @@
+
+		Schema: map[string]*schema.Schema{
+			consts.FieldName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of the destination.",
+				ForceNew:    true,
+			},
+			consts.FieldType: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Type of sync destination.",
+				ForceNew:    true,
+			},
+			consts.FieldMount: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the mount where the secret is located.",
+			},
+			fieldSecretName: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Specifies the name of the secret to synchronize.",
+			},
+			fieldSyncStatus: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Specifies the status of the association.",
+			},
+			fieldUpdatedAt: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Duration string stating when the secret was last updated.",
+			},
+		},
+	}
+}
+
+func secretsSyncAssociationWrite(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return diag.FromErr(e)
+	}
+
+	name := d.Get(consts.FieldName).(string)
+	destType := d.Get(consts.FieldType).(string)
+	secretName := d.Get(fieldSecretName).(string)
+	mount := d.Get(consts.FieldMount).(string)
+
+	path := secretsSyncAssociationSetPath(name, destType)
+
+	data := map[string]interface{}{
+		fieldSecretName:   secretName,
+		consts.FieldMount: mount,
+	}
+
+	log.Printf("[DEBUG] Writing association to %q", path)
+	resp, err := client.Logical().WriteWithContext(ctx, path, data)
+	if err != nil {
+		return diag.Errorf("error setting secrets sync association %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Wrote association to %q", path)
+
+	// expect accessor to be provided from mount
+	accessor, err := getMountAccessor(ctx, d, meta)
+	if err != nil {
+		return diag.Errorf("could not obtain accessor from given mount; err=%s", err)
+	}
+
+	// associations can't be read from Vault
+	// set data that is received from Vault upon writes
+	vaultRespKey := fmt.Sprintf("%s/%s", accessor, secretName)
+	associatedSecrets := "associated_secrets"
+	secretData, ok := resp.Data[associatedSecrets]
+	if !ok {
+		// did not find any associated secrets for this mount
+		// we expect to see data here since we just wrote an association
+		return diag.Errorf("expected associated secrets; received no secret associations in mount")
+	}
+
+	for _, k := range secretsSyncAssociationFields {
+		if secretsMap, ok := secretData.(map[string]interface{}); ok {
+			if val, ok := secretsMap[vaultRespKey]; ok {
+				if valMap, ok := val.(map[string]interface{}); ok {
+					if v, ok := valMap[k]; ok {
+						if err := d.Set(k, v); err != nil {
+							return diag.FromErr(err)
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
possible to remove the deep nesting here?

Reply:
Since the information we need is pretty nested within the [response returned from Vault] (https://developer.hashicorp.com/vault/api-docs/system/secrets-sync#sample-response-5) (`data/associated_secrets/kv_eb4acbae/my-secret-1`, for example), I'm not sure if we can avoid nesting here. I added type assertion checks at each step while doing so just to be safe

Would be curious to know if folks had thoughts on how we might be able to remove/simplify the nesting. Since it seems to be working robustly within the tests at the moment, I'll probably merge as is and can polish things like this in a follow-up PR if need be! 🙏🏼

Reply:
I like the struct idea. We could only add the fields we need for TFVP. If unmarshalling fails, we'd know that the API made breaking changes.

Reply:
This is the pattern we will use for new resources built with the Terraform Plugin Framework. See https://github.com/hashicorp/terraform-provider-vault/pull/2059/files#diff-a2228b67b947abd7a8f052592d74a00aabd42fc3186fd0140df2d95ed9c1b000R166-R171

Reply:
Thanks a lot for the great feedback folks! Added a struct model that we can unmarshal the Vault response into in order to avoid the nesting. Tests seem to work!